### PR TITLE
Eval refactor

### PR DESCRIPTION
**Pull request by shane-kercheval**
# Objective

The intent of this PR is to simplify a few areas of framework and address areas that needed improvement, either in the code quality or user interaction.

This PR contains breaking changes. The majority of these breaking changes will probably be minimal to current consumers of the library (e.g. llm-evaluations). For example, stead of `from_yaml` methods on evals/results/etc, the function has been changed to `from_file` which support yaml/json and can be extended in the future. Another example is "value_extractor" property in checks have been renamed to "data_path"; however, most checks don't set that property and won't have to be updated.

One change that will affect all users is that EvalHarnesss now returns a list of `CandidateRunResult` objects rather than a nested list of EvalResults. Previously, the inner list contained a list of EvalResults for each candidate and the outer list corresponded to the candidates. Now the interface is cleaner. CandidateRunResult contains information about about errors that were generated when running evals (eg. response errors or eval errors). It has an `eval_results` property  which correspond to the list EvalResults that was previously an inner list. Now, however the list will contain None entries for Evals that generated an error, and there are other properties to get those errors.

# Changes

- updated `ResponseData` -> `ResponseModel`
    - added `extract_values` function
- changed `Check` class so that TestHarness (or any code that runs many checks) can call a separate function for passing the ResponseModel object
    - users now implement the abstract method `__call__`
    - This means that users that use checks directly would do something like `precision = PrecisionScore(); precision(actual=..., expected=...)` rather than `precision(ResponseData(expected=, response=...)`.
    - In other words, the Eval doesn't know what data each check uses, so it calls this new function and passes in all of the data (i.e. `ResponseModel`) and the checks extract the data according to the `data_path` value on the check (which is a path (or multiple paths) into the ResponseModel).
- added auto-timestamp to CandidateResponse
- updated `get_value_from_path` in `internal_utilities.py` to be `singledispatch` and added support for dicts/lists
- refactored `to_yaml`/`from_yaml` out of classes to `SerializationMixin` and added `to_json`/`from_json` and `from_file` (which calls `from_json` or `from_yaml` according to extension)
    - Updated Eval/EvalResult/Candidate
    - Changed EvalHarness to use `add_eval_from_yaml` to `add_eval_from_file` (same with candidates.)
- deserializing EvalResult no longer instantiates Eval object or Candidate object, it simply uses the dictionaries.
    - trying to instantiate will cause continuous versioning issues, and if needed, the user can instantiate manually from dict
- decoupled Eval from Candidate
    - Eval class is now used directly against responses instead of candidates.
    - This is useful in many use-cases where responses already exist (e.g. complex response generation, rerunning evals with updated/fixed code, etc.)
    - EvalHarness still uses Candidates as a way to run many evals against many Evals, but now additional workflows can be created for processing a lot of evals.
    - Response generation could be decouple from evals so that a separate workflow, for example
        - asynchronously generates response; or, for example, 
        - llm-as-a-judge checks evals can be run asynchronously
        - cpu-intentsive evals (e.g. python code generation) can run in parallel
- Updated and improved EvalHarness async/multi-processing logic. 
    - Now separates sync/async/multiprocessing of `response generation` and `running evals` (e.g. can run response generation asynchronously and evals in parallel)
    - Improved overall code/stability
    - Improved testing over combinations of sync/async/multiprocessing
    - Introduce `EvalRunResult` dataclass that contains results and errors associated with response generation and running evals
    - Introduce `CandidateRunResult` which makes it convenient to iterate over eval results and check for errors, etc..
    - Batching asynchronous tasks via semaphore rather than complicated batching logic

### CODE REVIEW

File: examples/rag_evals_via_nvidia_chatrag_bench.py [hunk 1]
```diff
@@ -108,26 +108,26 @@ def __call__(self, input: dict) -> CandidateResponse:  # noqa: A002
...
-
 candidates = [CustomRAGAgentCandidate(**params) for params in param_combinations]
-harness = EvalHarness(
-    num_cpus=None,
-    async_batch_size=1,
-    num_samples=1,
-)
-harness.add_evals_from_yamls("examples/nvidia_chatrag_bench/evals/*.yaml")
+harness = EvalHarness(generate_mode=Mode.PARALLEL, eval_mode=Mode.PARALLEL)
+harness.add_evals_from_files("examples/nvidia_chatrag_bench/evals/*.yaml")
```
Comment (ahuang11) [inline examples/rag_evals_via_nvidia_chatrag_bench.py:113]: Long method names are clunky. I wonder if you can refactor it to `add_evals` and then if it's a path, use the internal method `_add_evals_from_files` else use the other method?

File: examples/rag_evals_via_nvidia_chatrag_bench.py [hunk 2]
```diff
@@ -108,26 +108,26 @@ def __call__(self, input: dict) -> CandidateResponse:  # noqa: A002
...
-    result.to_yaml(f"examples/nvidia_chatrag_bench/results/{result.eval_obj.metadata['uuid']}-{result.candidate_obj.metadata['uuid']}.yaml")
-
-
-harness.callback = save_callback
 print(f"Starting Evals: {len(harness.evals)} evals and {len(harness.candidates)} candidates (total {len(harness.evals) * len(harness.candidates)} eval-candidate pairs)")  # noqa: E501
-results = harness()
-print("Finished Evals")
+print(f"Timestamp: {time.strftime('%Y-%m-%d %H:%M:%S')}")
+results: list[CandidateRunResult] = harness()
+print(f"Finihsed - Timestamp: {time.strftime('%Y-%m-%d %H:%M:%S')}")
```
Comment (ahuang11) [inline examples/rag_evals_via_nvidia_chatrag_bench.py]:
```suggestion
print(f"Finished - Timestamp: {time.strftime('%Y-%m-%d %H:%M:%S')}")
```

Also, maybe in the future, incorporating logging through like logaru?

File: llm_eval/checks.py [hunk 3]
```diff
@@ -313,48 +285,63 @@ def _call(self, response: str, metadata: dict[str, Any]) -> CheckResult:
...
-            # if the value_extractor is a dictionary, the keys correspond to the names of the
-            # parameters in the Check subclass `_call` method and the values are the paths to the
-            # values in the ResponseData object; so we will pass as keyword arguments
-            result = self._call(**check_value)
-        else:
-            result = self._call(check_value)
-        if error or (self.value_extractor != self.default_value_extractor):
-            result.metadata['value_extractor'] = self.value_extractor
-            result.metadata['value_extracted'] = check_value
+    def via_response_model(self, data: ResponseModel) -> CheckResult:
```
Comment (ahuang11) [inline llm_eval/checks.py]: `via` to me, sounds like `from`, which sounds like a class method. 

`Check().apply_on_model()`? not sure

**Review by ahuang11 — APPROVED** [review]
Minor comments.

I haven't tested on Lumen, but I think these changes make sense to me and I don't mind the breaking changes; I really appreciate the `
Response generation could be decouple from evals so that a separate workflow` so looks good to me!

File: examples/rag_evals_via_nvidia_chatrag_bench.py (hunk 1 above)
Comment (shane-kercheval) [reply examples/rag_evals_via_nvidia_chatrag_bench.py:113]: Great idea, i will update the code.

File: llm_eval/checks.py (hunk 3 above)
Comment (shane-kercheval) [reply llm_eval/checks.py]: Great suggestion, i'll change!

**Pull request merged**